k8s: split containers so we have one process per container

alenkacz opened the pull request:
## Cover letter

We have a couple of customers that are seeing frequent OOM restarts of redpanda containers (which is causing problems to availability of the whole cluster). Right now it’s hard to say what is causing this mainly because there’s multiple processes running inside the container (which is in general not the best practice - ideally it should be one process per container). With multiple processes, we can never be sure which one is leaking memory causing OOM on the whole redpanda container.

Currently, this is running inside the container:
- WASM
- Call-home functionality
- Redpanda process (we call rpk start here)

The proposal is to introduce 2 separate containers with their own resources assigned:
- Redpanda
- Call home (rpk status)

WASM will be integrated separately as a feature for the operator. It does not make sense to run it right now either as a sidecar or as a process inside redpanda container.

[ch3591]

## Release notes

If the PR changes the user experience, write a short summary of the changes. See the [CONTRIBUTING](https://github.com/vectorizedio/redpanda/blob/dev/CONTRIBUTING.md) guidelines for details.

Release note: [1-2 sentences of what this PR changes]
Kubernetes operator now creates pods with multiple containers with one process per container to prevent OOMs. Updating to this version will cause restart of all nodes.

alenkacz commented on src/go/k8s/tests/e2e/external-connectivity/00-assert.yaml:
```diff
@@ -2,29 +2,6 @@ apiVersion: apps/v1
 kind: StatefulSet
 metadata:
   name: external-connectivity
```
this is covered by unit tests

alenkacz commented on src/go/k8s/tests/e2e/update-image-tls-client-auth/00-current-image.yaml:
```diff
@@ -4,7 +4,7 @@ metadata:
 spec:
   image: "vectorized/redpanda"
-  version: "v21.4.10"
+  version: "v21.6.6"
```
old images does not have the scripts we need

alenkacz commented on src/go/k8s/tests/e2e/update-conf-image/03-assert.yaml:
```diff
@@ -34,9 +29,5 @@ apiVersion: v1
 kind: Pod
 metadata:
   name: update-image-cluster-1
```
this is tested many other places

alenkacz commented on src/go/k8s/tests/e2e/produce-consume/00-assert.yaml:
```diff
@@ -2,29 +2,6 @@ apiVersion: apps/v1
 kind: StatefulSet
 metadata:
   name: cluster-sample
```
we have unit tests to cover these

RafalKorepta commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_webhook.go:
```diff
@@ -56,14 +82,14 @@ var _ webhook.Defaulter = &Cluster{}
 func allResourceFields(c *Cluster) []resourceField {
 	return []resourceField{
-		{c.Spec.Resources, field.NewPath("spec").Child("resources")},
+		{&c.Spec.Resources, field.NewPath("spec").Child("resources")},
```
Could this change be moved to previous commit? https://github.com/vectorizedio/redpanda/pull/1948/commits/884ad5d8bf82e034c04cc01eb044cdb51f4d6238

RafalKorepta commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_webhook.go:
```diff
@@ -48,7 +48,7 @@ var (
 		},
 	}
-	// DefaultWASMResources is default resources for coproc sidecar 
+	// DefaultWASMResources is default resources for coproc sidecar
```
Could this change be moved to previous commit? https://github.com/vectorizedio/redpanda/pull/1948/commits/67f1f0ac2b5672721ab3d70574d12ded7c4c1eac

RafalKorepta commented on src/go/k8s/pkg/resources/statefulset.go:
```diff
@@ -471,6 +444,82 @@ func (r *StatefulSetResource) obj() (k8sclient.Object, error) {
+
+func (r *StatefulSetResource) wasmContainer() corev1.Container {
+	if r.pandaCluster.Spec.DeepCopy().ResourcesWASM == nil {
+		r.logger.Info("BUG! No resources found for wasm - this should never happen with defaulting webhook enabled")
```
If this is a `BUG!` it should be at the error log level. right?

RafalKorepta replied:
Maybe message to the user should mention `please consider enabling defaulting webhook`?

alenkacz replied:
it could be bug or disabled webhook, also error requires error and I had nothing so I used info :)

RafalKorepta commented on src/go/k8s/pkg/resources/statefulset.go:
```diff
@@ -471,6 +444,82 @@ func (r *StatefulSetResource) obj() (k8sclient.Object, error) {
+		Name:      wasmContainerName,
+		Image:     r.pandaCluster.FullImageName(),
+		Command:   []string{"/usr/local/bin/coproc.sh"},
+		Env:       redpandaEnvVars(),
```
I don't understand what environment variable is needed for `coproc`? To be honest I don't know if any of the env vars are needed for Redpanda as all the heavy lifting is done by configurator process.

RafalKorepta commented on src/go/k8s/pkg/resources/statefulset.go:
```diff
@@ -471,6 +444,82 @@ func (r *StatefulSetResource) obj() (k8sclient.Object, error) {
+
+func (r *StatefulSetResource) rpkStatusContainer() corev1.Container {
+	if r.pandaCluster.Spec.ResourcesRpkStatus == nil {
+		r.logger.Info("BUG! No resources found for rpk status - this should never happen with defaulting webhook enabled")
```
ditto

RafalKorepta commented on src/go/k8s/pkg/resources/statefulset.go:
```diff
@@ -471,6 +444,82 @@ func (r *StatefulSetResource) obj() (k8sclient.Object, error) {
+		Name:      rpkStatusContainerName,
+		Image:     r.pandaCluster.FullImageName(),
+		Command:   []string{"/usr/local/bin/rpk-status.sh"},
+		Env:       redpandaEnvVars(),
```
I think only `REDPANDA_ENVIRONMENT` is necessary here, as we report this back to the `metrics` service.

REF: https://github.com/vectorizedio/redpanda/pull/567

RafalKorepta reviewed (changes requested):
Good job! I left few comments.

dimitriscruz commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_webhook.go:
```diff
@@ -308,59 +355,63 @@ func (r *Cluster) validatePandaproxyListeners() field.ErrorList {
-	// Ensure spare memory for other processes
-	quantity := resource.MustParse(ReserveMemoryString)
-	if !r.Spec.Configuration.DeveloperMode && (r.Spec.Resources.Requests.Memory().Value()-quantity.Value()) < gb {
+	if r.Spec.Resources.Requests.Memory().Value() < gb {
```
sure, the first one could have been removed when the latter was introduced

dimitriscruz replied:
We could keep the comment of the first check though, because the motivation is different.

alenkacz replied:
I adjusted, let me know if this works

senior7515 commented:
our cloud doesn't have wasm enabled by default (we should disable if so) - so we should make sure that wasm container is optional if wasm is off

dimitriscruz commented on src/go/k8s/pkg/resources/statefulset.go:
```diff
@@ -425,6 +395,8 @@ func (r *StatefulSetResource) obj() (k8sclient.Object, error) {
 								},
 							}, r.secretVolumeMounts()...),
 						},
+						r.rpkStatusContainer(),
+						r.wasmContainer(),
```
Looks like with this change we'd *always* be running wasm and the status container. What are the side-effects of this?

alenkacz replied:
yes this was the default right now, it was always disabled, for both. I will add an ability to disable them, though

alenkacz commented on src/go/k8s/apis/redpanda/v1alpha1/cluster_webhook_test.go:
```diff
@@ -179,7 +179,7 @@ func TestValidateUpdate(t *testing.T) {
+					corev1.ResourceMemory: resource.MustParse("0.9Gi"),
 				},
 			},
 		},
```
I added some tests, some things were already covered. Honestly the tests are mess, I would like to refactor that when this merges - I would like to split them into multiple methods and possibly share somehow between create and update as this is just very disorganized and hard to say what's tested and what's not

RafalKorepta replied:
Agree with you

alenkacz commented:
all: I addressed all comments and made changes to the overall feature:
- the sidecar can now be disabled (I think it's fair to expose it)
- WASM was dropped, we should add it as a feature when we're ready

RafalKorepta reviewed (approved):
LGTM

RafalKorepta reviewed (approved):
LGTM

alenkacz commented:
Honestly I feel very YOLO because I don't understand if it's safe to merge with 3 failing builds but ... it was safe last time 🤷‍♀️
